src/daemons/linuxd/src/worker_thread.rs

```diff
@@ -825,60 +823,78 @@ impl WorkerThreadHandle {
     }
 
     fn handle_read_request(
-        gateway_stdin_tx: Option<Sender<(ReadRequest, Sender<Message>)>>,
-        venv: Arc<Mutex<VirtualEnviromentDirectory>>,
+        gw_stream: Option<Arc<Mutex<SocketStream>>>,
         source: ThreadIdentifier,
         request: ReadRequest,
     ) -> Result<Message, WorkerThreadError> {
         trace!("handle_read_request(): source={source:?}, request={request:?}");
         // Check if reading from gateway.
         if request.fd == STDIN_FILENO {
-            let gateway_stdin_tx = if let Some(gateway_stdin_tx) = gateway_stdin_tx {
-                gateway_stdin_tx
+            let gw_stream = if let Some(gw_stream) = gw_stream {
+                gw_stream
             } else {
                 error!(
                     "handle_read_request(): process tried to read from stdin but no gateway found"
                 );
                 return Ok(ReadResponse::build(source, 0, [0u8; ReadResponse::BUFFER_SIZE]));
             };
 
-            // Check if the process is associated with a virtual environment.
-            let mut venv: MutexGuard<'_, VirtualEnviromentDirectory> = venv.lock().unwrap();
-            let env: &mut VirtualEnvironment = if let Some(env) = venv.get_mut(source) {
-                env
-            } else {
-                warn!(
-                    "handle_read_request(): process is not associated with a virtual environment, \
-                     returning EOF"
-                );
-                return Ok(ReadResponse::build(source, 0, [0u8; ReadResponse::BUFFER_SIZE]));
+            let response = {
+                // Take the lock (handle poison however you prefer)
+                let mut locked_gw_stream = match gw_stream.lock() {
+                    Ok(g) => g,
+                    Err(e) => {
+                        error!("gateway stream mutex poisoned (error={e:?})");
+                        return Ok(ReadResponse::build(
+                            source,
+                            0,
+                            [0u8; ReadResponse::BUFFER_SIZE],
+                        ));
```

Comment:
(nitpick) We may encapsulate the creation of a zero-sized message into something like `ReadResponse::eof()`.

Reply:
I can take care of this in a follow-up PR. I am creating a tracking issue: https://github.com/nanvix/nanvix/issues/772